### PR TITLE
Feature/animation refactor

### PR DESCRIPTION
**Pull request by nvh**
This is a full implementation of the proposal in #378

### CODE REVIEW

File: framer/Animation.coffee [hunk 1]
```diff
@@ -33,69 +33,48 @@ evaluateRelativeProperty = (target, k, v) ->
...
-# _runningAnimations = []
-
-# Todo: this would normally be BaseClass but the properties keyword
-# is not compatible and causes problems.
 class exports.Animation extends BaseClass
 
-	constructor: (options={}) ->
-
-		options = Defaults.getDefaults "Animation", options
+	constructor: (parameters={},options={}) ->
```
Comment (koenbok) [inline framer/Animation.coffee]: missing space

File: framer/Animation.coffee [hunk 2]
```diff
@@ -33,69 +33,48 @@ evaluateRelativeProperty = (target, k, v) ->
...
+		_.defaults(options, parameters.options)
 
-		super options
+		delete parameters.options
+		@options = _.cloneDeep Defaults.getDefaults "Animation", options
+		super parameters
+		@layer = parameters.layer ? null
```
Comment (koenbok) [inline framer/Animation.coffee]: We should throw an error if there is no layer here, and we should make the layer a readonly property.

File: framer/Animation.coffee [hunk 3]
```diff
@@ -33,69 +33,48 @@ evaluateRelativeProperty = (target, k, v) ->
...
 		set: (value) ->
 			@options?.looping = value
-			if @options?.looping and @options?.layer? and !@isAnimating
+			if @options?.looping and @layer? and !@isAnimating
 				@restart()
 
 	start: =>
-		if @options.layer is null
+		if @layer is null
 			console.error "Animation: missing layer"
```
Comment (koenbok) [inline framer/Animation.coffee]: If we check in constructor we can skip check here

File: framer/Layer.coffee [hunk 4]
```diff
@@ -883,35 +885,82 @@ class exports.Layer extends BaseClass
...
 
-	animate: (options) ->
-		# console.warn "Layer.animate is deprecated: please use Layer.animateTo instead"
-		properties = options.properties
-		delete options.properties
-		@animateTo(properties, options)
-
-	animateTo: (properties,options={}) ->
-		_.defaults(options,properties.options)
+	animateToState: (stateName, options={}) ->
```
Comment (koenbok) [inline framer/Layer.coffee]: This needs some more comments to make clear what exactly is happening.

File: framer/Layer.coffee [hunk 5]
```diff
@@ -883,35 +885,82 @@ class exports.Layer extends BaseClass
...
+			if not shouldChange
+				return null
+		finished = options.completion
+		options.completion = =>
+			@_stateMachine.emit(Events.StateDidSwitch, @_stateMachine.previousName, @_stateMachine.currentName, @)
+			finished?()
+		@animate properties, options
+
+	animate: (properties,options={}) ->
+		if typeof properties == 'string'
```
Comment (koenbok) [inline framer/Layer.coffee]: Style ' "

File: framer/Layer.coffee [hunk 6]
```diff
@@ -883,35 +885,82 @@ class exports.Layer extends BaseClass
...
-		delete options.properties
-		@animateTo(properties, options)
-
-	animateTo: (properties,options={}) ->
-		_.defaults(options,properties.options)
+	animateToState: (stateName, options={}) ->
+		properties = @_stateMachine.switchTo stateName
+		if @_stateMachine.previousName is @_stateMachine.currentName
+			shouldChange = false
+			for property,value of properties
```
Comment (koenbok) [inline framer/Layer.coffee]: Style: space

File: framer/Layer.coffee [hunk 7]
```diff
@@ -883,35 +885,82 @@ class exports.Layer extends BaseClass
...
+				if @[property] != value
```
Comment (koenbok) [inline framer/Layer.coffee]: Style: we prefer `isnt`

Comment (koenbok) [inline framer/Layer.coffee]: http://stackoverflow.com/questions/523643/difference-between-and-in-javascript

File: framer/Layer.coffee [hunk 8]
```diff
@@ -883,35 +885,82 @@ class exports.Layer extends BaseClass
...
+
+		animatableProperties = Animation.filterAnimatableProperties(properties)
+		nonAnimatableProperties = _.omit(_.clone(properties),_.keys(animatableProperties))
 
 		start = options.start
 		start ?= true
 		delete options.start
-
-		if options.instant
+		instant = options.instant ? false
```
Comment (koenbok) [inline framer/Layer.coffee]: http://stackoverflow.com/a/10146123/56495

**Comment by koenbok** [comment]
General some styling stuff:

- `'` vs `"`
- missing spaces (function arguments)
- `!=` vs `isnt` etc.

File: framer/Layer.coffee (hunk 8 above)
Comment (nvh) [inline framer/Layer.coffee]: It's not a ternary, but an existential operator: http://coffeescript.org/#operators

File: framer/Animation.coffee (hunk 2 above)
Comment (nvh) [inline framer/Animation.coffee]: No, because I'd like to be able to clone an animation and set its layer to some other layer before starting it

**Comment by nvh** [comment]
This is about finished and it could be merged in, but there are 3 breaking changes:

- The `default` state is renamed to `initial`
- `layer.states.current` has been renamed to `layer.states.currentName`
- `layer.states.previous` has been renamed to `layer.states.previousName`

Especially the last two are tricky, because `layer.states.current` and `layer.states.previous` do now contain an object with the specified state properties, so if you have the code:
```coffee
if layer.states.current == "stateA"
   print "The best state ever" # Will never happen
```
The code compiles, but doesn't do what you would expect with the previous API. We could show a warning or something, but that would hinder people using the new API as intended.

What do you think?

**Pull request merged**